File: b+tree.cpp
#include <bits/stdc++.h>
using namespace std;

int N = 3;
int leaf_keys = ceil((float)(N+1)/(float)2);
int non_leaf_keys = ceil((float)(N)/(float)2);     // size for non-leaf values split.
int non_leaf_ptrs = ceil((float)(N+2)/(float)2);
int maxx = 1e9+1;

bool DM = false;

typedef struct node{

    node* parent;
    vector<int> keys;
    vector<node*> pointers;
    int isLeaf;
    node* next_pointer;

}Node;

Node* init_node(int is_leaf,Node* parent)
{

    Node* node = new Node;
    node->parent = parent;
    node->keys = vector<int>(N,maxx);
    node->pointers = vector<Node*>(N+1);
    node->isLeaf = is_leaf;
    node->next_pointer = NULL;
    return node;
}

int find_lower_index(Node* node, int value)
{
    // finds the index which has a value just less than the given value.
    // bascially (lower_bound - 1)
    auto it = lower_bound(node->keys.begin(),node->keys.end(),value);
    int ind = it - node->keys.begin();
    return ind;
}

void insert(Node* node,int value,Node* new_node)
{
    // root is returned from the function
    // node in which we have to insert will be a leaf
    if(node->keys[N-1] == maxx)
    {
        // means there is still space in this leaf node
        // therefore find the first index which is empty, and then sort it.
        int ind = find_lower_index(node,value);
        node->keys.pop_back();
        node->pointers.pop_back();
        node->keys.insert(node->keys.begin()+ind,value);


        if (node->isLeaf != 1)
        {
            node->pointers.insert(node->pointers.begin()+ind+1,new_node);
        }
        
        cout << "Key with value " << value << " inserted sucessfully\n";
        // cout << node->keys.size() << node->keys[0] << endl;

        // for(int i=0;i<node->keys.size();i++)
        // {
        //     cout << i << " ";
        //     if(node->keys[i] != maxx)
        //     {
        //         cout << node->keys[i];
        //     }
        //     cout << endl;
        // }
    }
    else
    {
        // there is no space in leaf. so split the root - call the split function
        // since there is no space , we make a vector with the keys in temp and sort the values
        auto it = lower_bound(node->keys.begin(),node->keys.end(),value);
        int ind = it - node->keys.begin()+1;
        cout << "ind " << ind << endl;
        int up_value;

        vector<int> temp_vals;
        vector<Node*> temp_ptrs;

        temp_vals.push_back(value);
        for(int i=0;i<N;i++)
        {
            temp_vals.push_back(node->keys[i]);       
        }
        sort(temp_vals.begin(),temp_vals.end());

        for(int i=0;i<N+1;i++)
        {
            temp_ptrs.push_back(node->pointers[i]);       
            if(i == ind-1)
            {
                temp_ptrs.push_back(new_node);
            }
        }

        // for(int i=0;i<=N;i++)
        // {
        //     cout << temp_vals[i] << " ";
        // }
        // cout << '\n';

        Node* split_node1 = init_node(node->isLeaf,node->parent);
        Node* split_node2 = init_node(node->isLeaf,node->parent);

        //  first ciel(n+1/2) values in one and rest in other

        if (node->isLeaf != 1)
        {
            for(int i=0;i<non_leaf_keys;i++)
            {
                split_node1->keys[i] = temp_vals[i];
            }
            for(int i=non_leaf_keys;i<temp_vals.size();i++)
            {
                split_node2->keys[i-non_leaf_keys] = temp_vals[i];
            }

            for(int i=0;i<non_leaf_ptrs;i++)
            {
                split_node1->pointers[i] = temp_ptrs[i];
            }    
            for(int i=0;i<non_leaf_ptrs;i++)
            {
                split_node2->pointers[i-non_leaf_ptrs] = temp_ptrs[i];
            }

            up_value = split_node2->keys[0];
            split_node2->keys.erase(split_node2->keys.begin());
            
        }
        else
        {
            for(int i=0;i<leaf_keys;i++)
            {
                split_node1->keys[i] = temp_vals[i];
            }
            for(int i=leaf_keys;i<temp_vals.size();i++)
            {
                split_node2->keys[i-leaf_keys] = temp_vals[i];
            }
            
            up_value = split_node2->keys[0];

            // for(int i=0;i<split_node1->keys.size();i++)
            // {
            //     cout << split_node1->keys[i] << " ";
            // }
            // cout << endl;

            // for(int i=0;i<split_node2->keys.size();i++)
            // {
            //     cout << split_node2->keys[i] << " ";
            // }
            // cout << endl;
            
        }
        
        split_node2->next_pointer = split_node1->next_pointer;
        split_node1->next_pointer = split_node2;

        // cout << "SegNO\n";
        //  first ciel(n+2/2) values in one and rest in other   
        *node = *split_node1;
        *new_node = *split_node2;
        // cout << "SegNO1\n";
        // for(int i=0;i<node->keys.size();i++)
        //     {
        //         cout << node->keys[i] << " ";
        //     }
        //     cout << endl;

        //     for(int i=0;i<new_node->keys.size();i++)
        //     {
        //         cout << new_node->keys[i] << " ";
        //     }
        //     cout << endl;
        
        if (node->parent != NULL)
        {
            cout << "blah\n";
            insert(node->parent,up_value,new_node);
        }
        return;
    }
}

Node* search(Node* node,int value)
{
    if (node->isLeaf == 1)
    {
        if (binary_search(node->keys.begin(),node->keys.end(),value) == 1)
        {
            cout << "Element with value " << value << " exists in the tree\n";
        }
        else
        {
            cout << "Element with value " << value << " not found\n";
        }
    
        return node;   // returns node pointer where the value is stored.
    }
    else
    {
        // vector<int>::iterator it = upper_bound(node->keys.begin(),node->keys.end(),value);                 // auto it;
        auto it = upper_bound(node->keys.begin(),node->keys.end(),value);
        int ind = it - node->keys.begin();
        node = node->pointers[ind];
        search(node,value);
    }
}

Node* insert_val(Node* root,int value)
{
    /// Returns root of the tree
    Node* temp = search(root,value);
    // cout << temp->keys.size() << endl;
    // temp will be a leaf node.
    // what if temp returns null. can it return null? , yes if node is not found. , but we do not want null here and hence the search function never returns a null
    Node * new_node = init_node(1,NULL);
    insert(temp,value,new_node);

    if(temp->parent == NULL)
    {
        
    }
    
    // for(int i=0;i<temp->keys.size();i++)
    // {
    //     if(temp->keys[i] != maxx)
    //     {
    //         cout << temp->keys[i] << " ";
    //     }
    // }
    // cout << '\n';

    // for(int i=0;i<new_node->keys.size();i++)
    // {
    //     if(new_node->keys[i] != maxx)
    //     {
    //         cout << new_node->keys[i] << " ";
    //     }
    // }
    // cout << '\n';

    return root;
}

int main(int argc,char const *argv[])
{

    // Individually store kar sakti hai
    string query_type[4] = {"FIND","INSERT","COUNT","RANGE"};
    string filename;

    if (argc <= 1)
    {
        cout << "Filename not given\n";
    }
    else
    {
        Node* root = init_node(1,NULL);
        filename = argv[1];
        ifstream file(filename);
        string line;
        int value1,value2;
        while(getline(file,line))
        {
            
            if(line.find(query_type[0]) != string::npos)
            {
                // FIND
                istringstream (line.substr(5)) >> value1;
                // cout << query_type[0] << " " << value1 << '\n';
            }
            else if(line.find(query_type[1]) != string::npos)
            {
                // INSERT
                istringstream (line.substr(7)) >> value1;
                root = insert_val(root,value1);
                // cout << query_type[1] << " " << value1 << '\n';
            }
            else if(line.find(query_type[2]) != string::npos)
            {
                // COUNT
                istringstream (line.substr(6)) >> value1;
                // cout << query_type[2] << " " << value1 << '\n';
            }
            else if(line.find(query_type[3]) != string::npos)
            {
                // RANGE
                istringstream (line.substr(6)) >> value1 >> value2;
                // cout << query_type[3] << " " << value1  << " " <<  value2 << '\n';
            }
            else
            {
                cout << "Command Not Found\n";
            }
        }
    }
        
}

File: b+_tree.py
from __future__ import print_function
import sys
import time
import os
import numpy as np
import bisect

def print_tree(node):
    
    for pointers in node.pointers:
        if node.is_leaf == 0:
            print_tree(pointers)
        else:
            for key in node.keys:
                print(key,end = " ")


def get_query_leaf(node,value):
	num_keys = len(node.keys)-1
	if node.is_leaf == 0:
		if node.keys[0] >= value:
			return get_query_leaf(node.pointers[0],value)
		if node.keys[num_keys] < value:
			return get_query_leaf(node.pointers[num_keys],value)
		for idx in range(num_keys):
			if value <= node.keys[idx+1]:
				if value > node.keys[idx]:
					return get_query_leaf(node.pointers[idx+1],value)
	elif node.is_leaf == 1:
		return node

def get_keys(node,min_value,max_value):
    key_count = 0
    num_keys = len(node.keys)
    if (num_keys == 0):
        return None,0
    
    for i in range(num_keys):
        if node.keys[i] >= min_value and node.keys[i] <= max_value:
            key_count = key_count + 1
    
    if node.keys[num_keys-2] <= max_value:
        return node.next,key_count
    if node.keys[num_keys-2] > max_value:
        return None,key_count

def count_range(tree,min_value,max_value):
    count = 0
    leaf = get_query_leaf(tree.root,min_value)
    while leaf is not None:
        leaf,count_leaf = get_keys(leaf,min_value,max_value)
        count += count_leaf
    return count

def split(node):
    sib_node = Node()
    sib_node.is_leaf = node.is_leaf
    num_keys = len(node.keys)
    mid = num_keys // 2
    mid_val = node.keys[mid]

    if node.is_leaf == 1:
        sib_node.keys = node.keys[mid:]
        sib_node.pointers = node.pointers[mid:]
        sib_node.next = node.next

        node.keys = node.keys[:mid]
        node.pointers = node.pointers[:mid]
        node.next = sib_node
    else:
        sib_node.keys = node.keys[mid+1:]
        sib_node.pointers = node.pointers[mid+1:]
        node.keys = node.keys[:mid]
        node.pointers = node.pointers[:mid+1]
    
    return mid_val,sib_node

def check_overflow(node,N):
    num_keys = len(node.keys)
    if num_keys > N:
        ## if total number is greater than what it can hold then split.
        return 1
    else:
        ## just added at that index, no change to be made
        return 0

def mid_check(node,mid,newNode,N):
    if mid is None:
        return None,None
    else:
        insert_index = bisect.bisect(node.keys,mid)
        node.keys.insert(insert_index,mid)
        node.pointers.insert(insert_index+1,newNode)
        if len(node.keys) > N:
            ## if total number is greater than what it can hold then split.
            return split(node)
        else:
            ## just added at that index, no change to be made
            return None,None

class Node:
    def __init__(self):
        self.keys = []
        self.pointers = []
        self.is_leaf = 1
        self.next = None
    
class Bplus_Tree():

    def __init__(self,N):
        
        self.root = Node()
        self.keys = []
        self.pointers = []
        self.is_leaf = 1
        self.num = N
        self.next = None
    
    def insert(self,node,value):

        num_keys = len(node.keys)
        flag = 0
        if not node.is_leaf:
            for i in range(num_keys):
                if flag == 1:
                    ## element has been inserted
                    break
                
                if value < node.keys[i] and i == 0:
                    ## value to be inserted at first position
                    mid , newNode = self.insert(node.pointers[0],value)
                    flag = 1
                elif value >= node.keys[i] and i == num_keys-1:
                    ## insert at the last
                    mid , newNode = self.insert(node.pointers[-1],value)
                    flag = 1
                elif value < node.keys[i+1] and value >= node.keys[i]:
                    ## insert at the i+1 th position
                    mid , newNode = self.insert(node.pointers[i+1],value)
                    flag = 1

            
        else:
            ## if not leaf
            insert_index = bisect.bisect(node.keys,value)
            node.keys.insert(insert_index,value)
            node.pointers.insert(insert_index,value)
            if len(node.keys) > self.num:
                ## if total number is greater than what it can hold then split.
                return split(node)
            else:
                ## just added at that index, no change to be made
                return None,None

            
        ## will have to send self.num because class variable.    
        return mid_check(node,mid,newNode,self.num)

def split_node(node):
	if node.is_leaf == 1:
		newNode = Node()
		newNode.is_leaf = node.is_leaf
		n = len(node.keys)
		mid = n // 2
		mid_value = node.keys[mid]
		newNode.keys = node.keys[mid:]
		newNode.pointers = node.pointers[mid:]
		newNode.next = node.next
		node.keys = node.keys[:mid]
		node.pointers = node.pointers[:mid]
		node.next = newNode
	else:
		newNode = Node()
		newNode.is_leaf = node.is_leaf
		n = len(node.keys)
		mid = n // 2
		mid_value = node.keys[mid]
		newNode.keys = node.keys[mid+1:]
		newNode.pointers = node.pointers[mid+1:]
		node.keys = node.keys[:mid]
		node.pointers = node.pointers[:mid+1]
	
	return mid_value, newNode

def main():
    
    with open(file_path,'r') as file:
        for line in file:
            query = line.strip().split()
            
            
            if len(query) == 2:
                val1 = query[1]
            
            if len(query) == 3:
                val1 = query[1]
                val2 = query[2]
            
            if(query[0] == "INSERT"):
                ## insert val1 into tree
                # mid,newNode = tree.insert(tree.root,val1)
                mid,newNode = tree.insert(tree.root,val1)

                if mid is not None:
                    new_root = Node()
                    new_root.keys = [mid]
                    new_root.is_leaf = 0
                    new_root.pointers = [tree.root,newNode]
                    tree.root = new_root
                
                print("-------Tree------")
                print_tree(tree.root)
                print()
                print("-------Treeend------")

            
            elif(query[0] == "FIND"):
                ## find val1 in tree
                count = count_range(tree,val1,val1)
                if count == 0:
                    print("NO")
                else:
                    print("YES")
            elif(query[0] == "COUNT"):
                ## count of val1 
                count = count_range(tree,val1,val1)
                print(count)
            elif(query[0] == "RANGE"):
                ## number of elements in range val1 to val2
                count = count_range(tree,val1,val2)
                print(count)
            else:
                print("Invalid Query")
            
            
if __name__ == "__main__":

    if len(sys.argv) < 2:
        print("Usage: python3 b+tree.py 'path_to_input_file';")
        sys.exit(1)
    else:
        N = 3
        command=sys.argv
        file_path = command[1]
        tree = Bplus_Tree(N)
        main()
